python/spear/env.py

```diff
@@ -553,3 +553,96 @@ def _serialize_array(array, space, byte_order):
+            motor_torque_constant   = 1.0 / motor_velocity_constant;                # Motor torque constant in [rad/s/V]
+
+            battery_voltage         = self._config.OPENBOT.BATTERY_VOLTAGE         # Voltage of the battery powering the OpenBot [V]
+            control_dead_zone       = self._config.OPENBOT.CONTROL_DEAD_ZONE       # Absolute duty cycle (in the ACTION_SCALE range) below which a command does not produce any torque on the vehicle
+            motor_torque_max        = self._config.OPENBOT.MOTOR_TORQUE_MAX        # Motor maximal torque [N.m]
+            electrical_resistance   = self._config.OPENBOT.ELECTRICAL_RESISTANCE   # Motor winding electrical resistance [Ohms]
+
+            # Speed multiplier defined in the OpenBot to map a [-1,1] action to a suitable command to 
+            # be processed by the low-level microcontroller. For more details, feel free to check the
+            # "speedMultiplier" command in the OpenBot code.
+            # https://github.com/isl-org/OpenBot/blob/master/android/app/src/main/java/org/openbot/vehicle/Vehicle.java#L375
+            action_scale = self._config.OPENBOT.ACTION_SCALE
+
+            # Acquire the ground truth motor and wheel velocity for motor counter-electromotive force
+            # computation purposes (or alternatively friction computation purposes).
+
+            # The ground truth velocity of the robot wheels in [RPM]
+            wheel_rotation_speeds = self._wheel_rotation_speeds
+
+            # The ground truth rotation speed of the motors in [rad/s]
+            motor_speed = gear_ratio * wheel_rotation_speeds * np.pi / 30.0 # Expressed in [rad/s]
+
+            # Compute the counter electromotive force using the motor torque constant
+            counter_electromotive_force = motor_torque_constant * motor_speed; # Expressed in [V]
+
+            # The electrical current allowed to circulate in the motor is the result of the
+            # difference between the applied voltage and the counter electromotive force
+            motor_winding_current = ((battery_voltage * duty_cycle) - counter_electromotive_force) / electrical_resistance; # Expressed in [A]
+
+            # The torque is then obtained using the torque coefficient of the motor in [N.m]
+            motor_torque = motor_torque_constant * motor_winding_current
+
+            # Motor torque is saturated to match the motor limits
+            motor_torque = np.clip(motor_torque, -motor_torque_max, motor_torque_max)
+
+            # The torque applied to the robot wheels is finally computed accounting for the gear ratio
+            wheel_torques = gear_ratio * motor_torque
+
+            # Control dead zone at near-zero speed. This is a simplified but reliable way to deal with
+            # the friction behavior observed on the real vehicle in the low-speed/low-duty-cycle regime.
+            for i in range(len(duty_cycle)):
```

Comment:
no need for a for-loop here, do this with a numpy one-liner

Reply:
Updated it